core/src/toga/widgets/base.py

```diff
@@ -282,6 +282,8 @@ def refresh(self):
             # We can't compute a layout until we have a viewport
             if self._impl.viewport:
                 super().refresh(self._impl.viewport)
+                # Refreshing the layout means the viewport needs a redraw.
+                self._impl.viewport.make_dirty()
```

Comment:
In core `Widget.refresh`.

Reply:
This is a `make_dirty()` with no arguments, so it's marking the *container* as dirty. On GTK, this triggers a "queue_resize" event, which (eventually) causes the actual layout to occur. 

In the GTK context, this results in a redundant layout calculation (the one 2 lines above the highlighted line); however, this call is the one that is *required* on other platforms.

To me, this is part of the problem with "container/viewport" abstraction. The `refresh()` API entry point is defined on Node, where it requires a viewport definition; but a good portion of the implementation in toga-core replicates the definition from Node so that the viewport can be injected. 

The existence of `refresh()` on toga.core.Widget is also a source of confusion, with people using it as a public facing API in the hope of applying *any* style change. Admittedly, this *has* been needed due to various bugs; however, once we've fixed those bugs (and this PR and #1761 fixes a lot of those cases), manual calls to `refresh()` shouldn't be needed at all. 

A lot of what `refresh()` has historically been used for could actually be replaced with `make_dirty()` on the container. If the places that are currently calling `some_widget.refresh()` was replaced with `some_widget.container.make_dirty()`, that would remove the redundant layout in the GTK case, and other platforms could implement `make_dirty()` as required (e.g., immediately invoking layout on the root widget of the layout with the appropriate viewport on Cocoa).

Reply:
Looking at the example you've provided - this `make_dirty()` won't be needed, because any change on the text widget that marks the text widget as needing a rehint (e.g., changing the text) will also flag the container as dirty, causing a layout. This specific `mark_dirty()` is needed for the case where a layout is happening but there's no widget-specific rehint requested. One example would be marking the text widget as `style.flex=1`. That will affect the layout, but not the hinting of a specific widget. 

Again, this would be a situation where it would be preferable for the applicator to invoke `mark_dirty()` directly on the container, and have backends respond with a layout at the appropriate time (immediately on Cocoa; on a redraw for GTK).